File: ghrah-protocol/src/ghrah/protocol/payloads/workspace.py
# SPDX-FileCopyrightText: 2026 chenxya <chenxya@ghrah.org>
#
# SPDX-License-Identifier: Apache-2.0
"""Workspace 域载荷模型（管理命令 + 事件）。"""

from __future__ import annotations

from pydantic import BaseModel

# ─── Workspace 管理载荷模型 ───


class CreateWorkspacePayload(BaseModel):
    """create_workspace 命令载荷。"""

    project_id: str
    agent_id: str
    agent_name: str


class DestroyWorkspacePayload(BaseModel):
    """destroy_workspace 命令载荷。"""

    project_id: str
    agent_id: str
    agent_name: str


class WorkspaceSnapshotPayload(BaseModel):
    """workspace_snapshot 命令载荷。"""

    project_id: str
    agent_id: str
    agent_name: str
    message: str = ""


class WorkspaceRollbackPayload(BaseModel):
    """workspace_rollback 命令载荷。"""

    project_id: str
    agent_id: str
    agent_name: str
    snapshot_id: str


class WorkspaceDiffPayload(BaseModel):
    """workspace_diff 命令载荷。"""

    project_id: str
    agent_id: str
    agent_name: str
    snapshot_id: str | None = None


class WorkspaceStatusPayload(BaseModel):
    """workspace_status 命令载荷。"""

    project_id: str
    agent_id: str
    agent_name: str


class WorkspaceRegisterPayload(BaseModel):
    """workspace_register 命令载荷（把已有目录登记为 workspace）。

    provider_type 为 None 时由 Subject 按 registry.detect 探测规则分派
    （有 .git → git；否则 → plain）。locator 为 file:// URI（MVP）。
    """

    locator: str
    name: str = ""
    provider_type: str | None = None


class WorkspaceGetPayload(BaseModel):
    """workspace_get 命令载荷（workspace_id 键控）。"""

    workspace_id: str


class WorkspaceListPayload(BaseModel):
    """workspace_list 命令载荷（可选 provider_type 过滤）。"""

    provider_type: str | None = None  # ─── Workspace 事件载荷模型 ───


class WorkspaceCreatedPayload(BaseModel):
    """workspace_created 事件载荷。"""

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    agent_name: str
    path: str


class WorkspaceDestroyedPayload(BaseModel):
    """workspace_destroyed 事件载荷。"""

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    agent_name: str


class WorkspaceSnapshotCreatedPayload(BaseModel):
    """workspace_snapshot_created 事件载荷。"""

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    agent_name: str
    snapshot_id: str
    message: str = ""


class WorkspaceRolledBackPayload(BaseModel):
    """workspace_rolled_back 事件载荷。"""

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    agent_name: str
    snapshot_id: str

File: ghrah-observer-webui/packages/protocol/src/payloads/workspace.ts
import { z } from "zod";

export const CreateWorkspacePayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  agent_name: z.string(),
});

export const DestroyWorkspacePayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  agent_name: z.string(),
});

export const WorkspaceSnapshotPayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  agent_name: z.string(),
  message: z.string().optional().default(""),
});

export const WorkspaceRollbackPayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  agent_name: z.string(),
  snapshot_id: z.string(),
});

export const WorkspaceDiffPayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  agent_name: z.string(),
  snapshot_id: z.string().nullable().optional(),
});

export const WorkspaceStatusPayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  agent_name: z.string(),
});

export const WorkspaceRegisterPayloadSchema = z.object({
  locator: z.string(),
  name: z.string().optional().default(""),
  provider_type: z.string().nullable().optional(),
});

export const WorkspaceGetPayloadSchema = z.object({
  workspace_id: z.string(),
});

export const WorkspaceListPayloadSchema = z.object({
  provider_type: z.string().nullable().optional(),
});

export const WorkspaceCreatedPayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  agent_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  agent_name: z.string(),
  path: z.string(),
});

export const WorkspaceDestroyedPayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  agent_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  agent_name: z.string(),
});

export const WorkspaceSnapshotCreatedPayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  agent_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  agent_name: z.string(),
  snapshot_id: z.string(),
  message: z.string().optional().default(""),
});

export const WorkspaceRolledBackPayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  agent_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  agent_name: z.string(),
  snapshot_id: z.string(),
});

export type CreateWorkspacePayload = z.infer<typeof CreateWorkspacePayloadSchema>;
export type DestroyWorkspacePayload = z.infer<typeof DestroyWorkspacePayloadSchema>;
export type WorkspaceSnapshotPayload = z.infer<typeof WorkspaceSnapshotPayloadSchema>;
export type WorkspaceRollbackPayload = z.infer<typeof WorkspaceRollbackPayloadSchema>;
export type WorkspaceDiffPayload = z.infer<typeof WorkspaceDiffPayloadSchema>;
export type WorkspaceStatusPayload = z.infer<typeof WorkspaceStatusPayloadSchema>;
export type WorkspaceRegisterPayload = z.infer<typeof WorkspaceRegisterPayloadSchema>;
export type WorkspaceGetPayload = z.infer<typeof WorkspaceGetPayloadSchema>;
export type WorkspaceListPayload = z.infer<typeof WorkspaceListPayloadSchema>;

export type WorkspaceCreatedPayload = z.infer<typeof WorkspaceCreatedPayloadSchema>;
export type WorkspaceDestroyedPayload = z.infer<typeof WorkspaceDestroyedPayloadSchema>;
export type WorkspaceSnapshotCreatedPayload = z.infer<typeof WorkspaceSnapshotCreatedPayloadSchema>;
export type WorkspaceRolledBackPayload = z.infer<typeof WorkspaceRolledBackPayloadSchema>;
